File: adapters/common/src/nemo_fabric_adapters/common/relay_artifacts.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Readiness checks for artifacts written asynchronously by NeMo Relay."""

from __future__ import annotations

import asyncio
import json
from collections.abc import Collection
from pathlib import Path
from typing import Any

from nemo_fabric_adapters.common import utils as common_utils

ATIF_FINALIZATION_TIMEOUT_SECONDS = 5.0
ATIF_POLL_INTERVAL_SECONDS = 0.05


def expects_local_atif(plugin_config: dict[str, Any]) -> bool:
    """Return whether Relay is configured to write ATIF to the local runtime.

    Relay treats a non-empty storage list as remote-only, so there is no local
    artifact for an adapter to await in that configuration.
    """

    for component in plugin_config.get("components", []):
        if (
            not isinstance(component, dict)
            or component.get("kind") != "observability"
            or component.get("enabled", True) is False
        ):
            continue
        config = component.get("config")
        if not isinstance(config, dict):
            continue
        atif = config.get("atif")
        if isinstance(atif, dict) and atif.get("enabled") and not atif.get("storage"):
            return True
    return False


def snapshot_atif_paths(plugin_config: dict[str, Any]) -> frozenset[Path]:
    """Capture ATIF paths that existed before an adapter invocation.

    Relay requires ``{session_id}`` in the ATIF filename template and creates a
    new session scope for each turn. Existing paths therefore cannot satisfy a
    later invocation, even if another process modifies them.
    """

    return frozenset(
        Path(artifact["path"])
        for artifact in common_utils.collect_relay_artifacts(plugin_config)
        if artifact.get("kind") == "atif"
    )


def _finalized_atif_path(
    plugin_config: dict[str, Any], before: Collection[Path]
) -> Path | None:
    """Find a newly created ATIF path containing a complete JSON object."""

    current = snapshot_atif_paths(plugin_config)
    for path in sorted(current.difference(before)):
        try:
            # Relay writes directly to the final path, so existence alone does
            # not prove that the JSON payload has been written completely.
            document = json.loads(path.read_bytes())
        except (OSError, UnicodeDecodeError, json.JSONDecodeError):
            continue
        if isinstance(document, dict):
            return path
    return None


async def wait_for_finalized_atif(
    plugin_config: dict[str, Any],
    before: Collection[Path],
    *,
    timeout_seconds: float = ATIF_FINALIZATION_TIMEOUT_SECONDS,
    poll_interval_seconds: float = ATIF_POLL_INTERVAL_SECONDS,
) -> Path | None:
    """Wait for one new, complete ATIF file until a monotonic deadline."""

    loop = asyncio.get_running_loop()
    deadline = loop.time() + timeout_seconds
    while True:
        if path := _finalized_atif_path(plugin_config, before):
            return path
        remaining = deadline - loop.time()
        if remaining <= 0:
            return None
        await asyncio.sleep(min(poll_interval_seconds, remaining))

File: adapters/typescript/pi/src/relay-artifacts.ts
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import { readFile, stat } from "node:fs/promises";
import { setTimeout as delay } from "node:timers/promises";

import { collectRelayArtifacts, type RelayPluginConfig } from "./relay-config.js";

export const ATIF_FINALIZATION_TIMEOUT_MS = 5_000;
export const ATIF_POLL_INTERVAL_MS = 50;

export type AtifSnapshot = Map<string, string>;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function pluginComponents(pluginConfig: RelayPluginConfig): unknown[] {
  return Array.isArray(pluginConfig.components) ? pluginConfig.components : [];
}

export function expectsLocalAtif(pluginConfig: RelayPluginConfig): boolean {
  for (const component of pluginComponents(pluginConfig)) {
    if (
      !isRecord(component) ||
      component.kind !== "observability" ||
      component.enabled === false ||
      !isRecord(component.config)
    ) {
      continue;
    }
    const atif = component.config.atif;
    if (isRecord(atif) && atif.enabled === true && (!Array.isArray(atif.storage) || atif.storage.length === 0)) {
      return true;
    }
  }
  return false;
}

async function fingerprint(path: string): Promise<string | undefined> {
  try {
    const value = await stat(path, { bigint: true });
    return `${value.dev}:${value.ino}:${value.size}:${value.mtimeNs}`;
  } catch {
    return undefined;
  }
}

export async function snapshotAtifFiles(pluginConfig: RelayPluginConfig): Promise<AtifSnapshot> {
  const snapshot: AtifSnapshot = new Map();
  for (const artifact of await collectRelayArtifacts(pluginConfig)) {
    if (artifact.kind !== "atif") {
      continue;
    }
    const value = await fingerprint(artifact.path);
    if (value !== undefined) {
      snapshot.set(artifact.path, value);
    }
  }
  return snapshot;
}

async function finalizedAtifPath(pluginConfig: RelayPluginConfig, before: AtifSnapshot): Promise<string | undefined> {
  const current = await snapshotAtifFiles(pluginConfig);
  for (const path of [...current.keys()].sort()) {
    if (before.get(path) === current.get(path)) {
      continue;
    }
    try {
      const document: unknown = JSON.parse(await readFile(path, "utf8"));
      if (isRecord(document)) {
        return path;
      }
    } catch {
      // Relay writes directly to the final path, so retry partial JSON.
    }
  }
  return undefined;
}

export async function waitForFinalizedAtif(
  pluginConfig: RelayPluginConfig,
  before: AtifSnapshot,
  options: { timeoutMs?: number; pollIntervalMs?: number } = {},
): Promise<string | undefined> {
  const timeoutMs = options.timeoutMs ?? ATIF_FINALIZATION_TIMEOUT_MS;
  const pollIntervalMs = options.pollIntervalMs ?? ATIF_POLL_INTERVAL_MS;
  const deadline = performance.now() + timeoutMs;
  while (true) {
    const path = await finalizedAtifPath(pluginConfig, before);
    if (path !== undefined) {
      return path;
    }
    const remaining = deadline - performance.now();
    if (remaining <= 0) {
      return undefined;
    }
    await delay(Math.min(pollIntervalMs, remaining));
  }
}
